server/src/main/java/org/opensearch/index/engine/NRTReplicationEngineFactory.java

```diff
@@ -8,18 +8,36 @@
 
 package org.opensearch.index.engine;
 
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.cluster.service.ClusterService;
+
 /**
  * Engine Factory implementation used with Segment Replication that wires up replica shards with an ${@link NRTReplicationEngine}
  * and primary with an ${@link InternalEngine}
  *
  * @opensearch.internal
  */
 public class NRTReplicationEngineFactory implements EngineFactory {
+
+    private final ClusterService clusterService;
+
+    public NRTReplicationEngineFactory(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    public NRTReplicationEngineFactory() {
+        this.clusterService = null;
+    }
```

Comment:
Should ClusterService really be optional?

Reply:
I'd made it optional for tests but I agree, makes more sense to just have 1 constructor with mandatory argument. 
If we don't have the reference to clusterService then whenever there's a clusterState change, we'll need to call another cluster state listener to update the node version values that we have in the engine that'll add a lot of unwanted traffic.